### ISSUE TITLE
create_resources requiring uniqueness for resource_source names across projects

### ISSUE THREAD

**Issue by wyardley** [voxpupuli/puppet-rundeck#231]
tldr; I want to be able to have the same resource source filename within multiple projects.

Given the following manifest:

```
  rundeck::config::project { 'test_project':
    file_copier_provider   => 'script-copy',
    node_executor_provider => 'script-exec',
    resource_sources => {
      'resources' => {
        'number' => '1',
        'include_server_node' => true,
        'resource_format' => 'resourceyaml'
      },
      'test2' => {
        'number' => '2',
        'include_server_node' => true,
        'resource_format' => 'resourceyaml'
      }
    }
  }

  rundeck::config::project { 'test_project2':
    file_copier_provider   => 'script-copy',
    node_executor_provider => 'script-exec',
    resource_sources       => { 'resources' => {
        'number' => '1',
        'include_server_node' => true,
        'resource_format' => 'resourceyaml'
      }
    }
  }
```

Puppet returns the error:

```
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Duplicate declaration: Rundeck::Config::Resource_source[resources] is already declared; cannot redeclare at /etc/puppet/modules/rundeck/manifests/config/project.pp:135 on node xxx.local
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
```

I realize this is a puppet constraint of uniqueness across resource names, regardless of scope, but seems like there might be a couple ways to work around this. I'm trying out a couple options to hack this feature into the module (simplest way would be to make sure that the resource $name has the project name prepended to it, I guess, or else to add an optional $file_name parameter, similar to $script_name, that would get set to $name if not explicitly set). Any ideas on a simpler way to work around this?

**Comment by rnelson0** [comment]
Can you share or link to the code on line 135 where the error is generated?
It's not clear what is being created here.

On Thursday, May 19, 2016, Will Yardley <notifications@github.com> wrote:

> tldr; I want to be able to have the same resource source filename within
> multiple projects.
>
> Given the following manifest:
>
>   rundeck::config::project { 'test_project':
>     file_copier_provider   => 'script-copy',
>     node_executor_provider => 'script-exec',
>     resource_sources => {
>       'resources' => {
>         'number' => '1',
>         'include_server_node' => true,
>         'resource_format' => 'resourceyaml'
>       },
>       'test2' => {
>         'number' => '2',
>         'include_server_node' => true,
>         'resource_format' => 'resourceyaml'
>       }
>     }
>   }
>
>   rundeck::config::project { 'test_project2':
>     file_copier_provider   => 'script-copy',
>     node_executor_provider => 'script-exec',
>     resource_sources       => { 'resources' => {
>         'number' => '1',
>         'include_server_node' => true,
>         'resource_format' => 'resourceyaml'
>       }
>     }
>   }
>
> Puppet returns the error:
>
> Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Duplicate declaration: Rundeck::Config::Resource_source[resources] is already declared; cannot redeclare at /etc/puppet/modules/rundeck/manifests/config/project.pp:135 on node c7node0.dev.axsops.local
> Warning: Not using cache on failed catalog
> Error: Could not retrieve catalog; skipping run
>
> I realize this is a puppet constraint of uniqueness across resource names,
> regardless of scope, but seems like there might be a couple ways to work
> around this. I'm trying out a couple options to hack this feature into the
> module (simplest way would be to make sure that the resource $name has the
> project name prepended to it, I guess, or else to add an optional
> $file_name parameter, similar to $script_name, that would get set to $name
> if not explicitly set). Any ideas on a simpler way to work around this?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/voxpupuli/puppet-rundeck/issues/231>
>


-- 

Rob Nelson
rnelson0@gmail.com

**Comment by rnelson0** [comment]
Crap, never mind, I though I was responding to a puppet-users email, and mobile doesn't grant me an edit button. #winning

**Comment by wyardley** [comment]
No worries, but yeah,  /etc/puppet/modules/rundeck/ is just this module.
We are doing more complicated stuff in our normal use case (and using an additional create_resources() with data from hiera), but this test case just has the class included and the above stanza in an enclosing role as a simple way of testing / reproducing the problem.

I was thinking about making a function to rename the key of the $resource_sources hash to $project-$resource and then renaming it back, but haven't been having much luck so far.

**Comment by jyaworski** [comment]
@wyardley if we really update stdlib here, we can use ensure_resources.

https://github.com/puppetlabs/puppetlabs-stdlib#ensure_resources

That doesn't completely solve your problem, but it may help.

**Comment by wyardley** [comment]
@jyaworski I don't think ensure_resources will do what we want here, unless it will overwrite / update the resource. From what I can see, it only creates the resource. But the problem here is the requirement for resource keys to be unique at global scope, but in this case, we want to have a resource with the same key (but potentially different values) within the scope of a particular project.

**Comment by jyaworski** [comment]
@wyardley it ensures a resource with those parameters exists in the catalog. If it does not, it creates it.

**Comment by wyardley** [comment]
@jyaworski right, but in this case, the resource for resource_sources can exist with the same name in different projects. So will it work correctly for the scenario outlined in the manifest above, where two or more projects each have a resource named, e.g., 'resources', potentially with different parameters? My understanding is that resources in Rundeck::Config::Resource_source have to be unique, so I'm not sure the behavior will be predictable in this case.

(I don't believe our stdlib has ensure_resources, and we're still on Puppet 3.8.x, so not sure if there's a version which does support it). With the newer features in Puppet 4.x (easier iteration etc.), I think it would be possible to support this more easily in another way, but have been stumped so far.

**Comment by wyardley** [comment]
It seems like it does a merge if the resource already exists (which I think in this case could cause unexpected or incorrect results).

**Comment by jyaworski** [comment]
@wyardley stdlib 4.x is compatible with puppet 3. ensure_resources was introduced in one of the newer releases.

**Comment by lgallard** [comment]
I have the same issue, but I'm using the hieradata definition as follow:

```
rundeck::projects:
  Project1:
    resource_sources:
      resources:
        project_name: 'Project1'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'

  Project2:
        project_name: 'Project2'
        number: '2'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
```


Is there anything wrong with this definition? Did I miss something?

**Comment by dalisch** [comment]
You almost had it - the resources are scoped to each project so you have to have a separate hash for each of them:

```
rundeck::projects:
  Project1:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project1'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
  Project2:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project2'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
```

**Comment by lgallard** [comment]
I put your suggestion but I got a duplicate declaration error.

hieradata:

```rundeck::projects:
  Project1:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project1'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
  Project2:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project2'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
```

Error:


> Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Duplicate declaration: Rundeck::Config::Resource_source[some_resorce_source_1] is already declared; cannot redeclare at...

 I also tried by modifying the hash like this:

```rundeck::projects:
  Project1:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project1'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
  Project2:
    resource_sources:
      some_resorce_source_2:
        project_name: 'Project2'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
```

But now I got this error:

> Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Resource Statement, Duplicate declaration: Ini_setting[resources.source.1.config.mappingParams] is already declared in file /etc/puppetlabs/code/environments/prd/modules/rundeck/manifests/config/resource_source.pp:301; cannot redeclare at 

And finally I also shanged the resourc number to 2:

```rundeck::projects:
  Project1:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project1'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
  Project2:
    resource_sources:
      some_resorce_source_2:
        project_name: 'Project2'
        number: '2'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
```

And it stopped complaining but the Project2 has not resource assigned. I guess due to the lack of a resource 1.

**Comment by lgallard** [comment]
Setting the projects with different resource hashes (I guess this is what @dalisch suggested originally):

```rundeck::projects:
  Project1:
    resource_sources:
      some_resorce_source_1:
        project_name: 'Project1'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
  Project2:
    resource_sources:
      some_resorce_source_2:
        project_name: 'Project2'
        number: '1'
        include_server_node: true
        resource_format: 'resourcexml'
        source_type: aws-ec2
        use_default_mapping: true
        mapping_params: 'tag.stopping.selector=;ami_id.selector=imageId;hostname.selector=privateIpAddress;username.default=rundeck'
```

And by fixing the resource_source.pp code as shown:

```
'aws-ec2': {
  ini_setting { "${name}::resources.source.${number}.config.mappingParams":
```

Because  there wasn't  name as in the file resource for aws-ec2.  Here is what I fixed:

https://github.com/lgallard/puppet-rundeck/blob/5290607667d4436a17eea8c9879b20bc3d736844/manifests/config/resource_source.pp#L301

And a PR against my master (it includes two fixes for AWS environments)

https://github.com/voxpupuli/puppet-rundeck/pull/293

I hope it helps out others setting projects using aws-ec2 plugin for resource source model.

**Comment by wyardley** [comment]
We've worked around this in a different way, so I'm going to (finally) close this one out.
